src/corecel/sys/KernelRegistry.hh

```diff
@@ -0,0 +1,126 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/sys/KernelRegistry.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <atomic>
+#include <iosfwd>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include "corecel/Assert.hh"
+#include "corecel/OpaqueId.hh"
+
+#include "KernelAttributes.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+struct KernelProfiling
+{
+    //!< Number of times launched
+    std::atomic<int> num_launches{0};
+    //!< Number of threads integrated over all launches
+    std::atomic<long long> accum_threads{0};
+
+    // Increment atomic counters given the number of threads
+    inline void log_launch(int num_threads);
+};
+
+//---------------------------------------------------------------------------//
+struct KernelMetadata
+{
+    std::string      name;
+    KernelAttributes attributes;
+    KernelProfiling  profiling;
+};
+
+//! Ordered identifiers for registered kernels
+using KernelId = OpaqueId<KernelMetadata>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Keep track of kernels and launches.
+ *
+ * This class has a thread-safe *insert* method because it's meant to be shared
+ * across multiple threads when running. Since accessing by ID is expected to
+ * be done well after all insertions have been completed, the \c size and \c at
+ * accessors are *not* thread safe.
+ */
+class KernelRegistry
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using value_type      = KernelMetadata;
+    using const_reference = const KernelMetadata&;
+    using key_type        = KernelId;
+    using size_type       = KernelId::size_type;
+    //!@}
+
+  public:
+    // Whether profiling metrics (launch count, max threads) are collected
+    static bool profiling();
+
+    // Construct without any data
+    KernelRegistry() = default;
+
+    // Register a kernel and return optional reference to profiling info
+    KernelProfiling* insert(const char* name, KernelAttributes&& attrs);
+
+    //! Number of kernel diagnostics available
+    size_type size() const { return kernels_.size(); }
```

Comment:
Technically this access should be protected by the lock.  Alternatively we could have a shadow copy of the size held in an atomic.

Reply:
Yeah, I wrote in the class description that these can only be accessed after all insertion has completed across all threads. I don't think adding a lock here would help us much, since if we were inserting at the same time our size would be less at the beginning of the accesses than at the end (i.e. we're not looping through the whole container).